src/encoding/path.rs

```diff
@@ -0,0 +1,368 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Also licensed under MIT license, at your choice.
+
+use bytemuck::{Pod, Zeroable};
+use peniko::kurbo::Shape;
+
+use super::Monoid;
+
+/// Path segment.
+#[derive(Clone, Copy, Debug, Zeroable, Pod)]
+#[repr(C)]
+pub struct PathSegment {
+    pub origin: [f32; 2],
+    pub delta: [f32; 2],
+    pub y_edge: f32,
+    pub next: u32,
+}
+
+/// Path segment type.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Pod, Zeroable)]
+#[repr(C)]
+pub struct PathSegmentType(pub u8);
+
+impl PathSegmentType {
+    /// Line segment.
+    pub const LINE_TO: Self = Self(0x1);
+
+    /// Quadratic segment.
+    pub const QUAD_TO: Self = Self(0x2);
+
+    /// Cubic segment.
+    pub const CUBIC_TO: Self = Self(0x3);
+}
+
+/// Path tag representation.
+#[derive(Copy, Clone, PartialEq, Eq, Pod, Zeroable)]
+#[repr(C)]
+pub struct PathTag(pub u8);
+
+impl PathTag {
+    /// 32-bit floating point line segment.
+    pub const LINE_TO_F32: Self = Self(0x9);
```

Comment:
As an outsider, a comment explaining that `0x9 = 0b1001 = Self::F32_BIT | PathSegmentType::LINE_TO` would help me understand these seemingly magic numbers.

Reply:
Good point. I'll add a comment.